prover/src/lib.rs

```diff
@@ -211,39 +300,204 @@ impl SP1Prover {
         &self,
         vk: &SP1VerifyingKey,
         sp1_challenger: Challenger<CoreSC>,
-        mut proofs: Vec<SP1ReduceProofWrapper>,
+        proofs: Vec<SP1ReduceProofWrapper>,
+        deferred_proofs: Vec<ShardProof<InnerSC>>,
         batch_size: usize,
     ) -> Vec<SP1ReduceProofWrapper> {
-        // If there's one proof at the end, push it to the next layer.
-        let last_proof = if proofs.len() % batch_size == 1 {
-            Some(proofs.pop().unwrap())
-        } else {
-            None
-        };
+        let last_proof = None;
 
-        // Process at most 4 proofs at once in parallel, due to memory limits.
+        // If there are deferred proofs, we want to add them to the end.
+        // OPT: If there's only one proof in the last batch, we could push it to the next layer.
+        // OPT: We could pack deferred proofs into the last chunk if it has less than batch_size proofs.
         let chunks: Vec<_> = proofs.chunks(batch_size).collect();
+
+        let mut reconstruct_challenger = self.setup_initial_core_challenger(vk);
+        let reconstruct_challengers = chunks
+            .iter()
+            .map(|proofs| {
+                let start_challenger = reconstruct_challenger.clone();
+                for proof in proofs.iter() {
+                    match proof {
+                        SP1ReduceProofWrapper::Core(reduce_proof) => {
+                            reconstruct_challenger
+                                .observe(reduce_proof.proof.commitment.main_commit);
+                            reconstruct_challenger.observe_slice(
+                                &reduce_proof.proof.public_values.to_vec()
+                                    [0..self.core_machine.num_pv_elts()],
+                            );
+                            let pv = PublicValues::<Word<Val<CoreSC>>, Val<CoreSC>>::from_vec(
+                                reduce_proof.proof.public_values.clone(),
+                            );
+                            println!("next_pc = {:?}", pv.next_pc);
+                            println!("shard = {:?}", pv.shard);
+                            println!("pv_digest = {:?}", pv.committed_value_digest);
+                        }
+                        SP1ReduceProofWrapper::Recursive(reduce_proof) => {
+                            let pv = RecursionPublicValues::from_vec(
+                                reduce_proof.proof.public_values.clone(),
+                            );
+                            reconstruct_challenger.sponge_state =
+                                pv.end_reconstruct_challenger.sponge_state;
+                            reconstruct_challenger.input_buffer =
+                                pv.end_reconstruct_challenger.input_buffer[..pv
```

Comment:
for here, seems like it's worth having some method to be like Challenger::from(...)